Spm/44 prepare new release

spmeisburger opened the pull request:
This is kind of a biggie.  We've been processing data with the `spm/23-parallel-write-data` branch.  Here, I've made some small changes relative to that branch that will allow us to merge into main:

1. Bumped the version to `1.0.3`
2. Updated the `README.md` files
3. Edited the `examples` to work with current version of the code, and re-ran the jupyter notebooks
4. Made `--datastore datastore` the default for `mdx2.import_data`
5. Removed some legacy code (dependencies on multiprocessing library, and alternative run methods for `nproc=1`)
6. Rebased on main (ready to merge).

This PR resolves issues #19, #23, #24, #29, #30, #40, and #44

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * New reintegrate command; many CLIs now use parameter objects, structured logging, and automatic parallelism. Virtual-dataset image saving and a 3D HKL grid type added.

* **Documentation**
  * README reworked (installation, examples, versions); examples and tutorial notebooks updated; DOI badge commented and fork banner added.

* **Bug Fixes / Security**
  * Safer Nexus I/O with version warnings, virtual-field fixes, and load-time security checks.

* **Tests**
  * Added tests for CLI parsing, virtual datasets, and Nexus load/save security.

* **Chores**
  * sandbox/ ignored; logging/parsing deps added.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

spmeisburger committed:
refactor to enable testing argparse interface

spmeisburger committed:
patched weird argparse behavior with boolean arguments

spmeisburger committed:
simple use case tests for CLI argument parsing

spmeisburger committed:
raise SystemExit on parse error

spmeisburger committed:
more elatorate test cases for CLI parsing

spmeisburger committed:
added simple-parsing dependency

spmeisburger committed:
represent parameters using dataclasses, configure CLI via simple-parsing

spmeisburger committed:
modify tests to use nested namespace variable names, absorption.enable instead of absorption_enable

spmeisburger committed:
implement 3d Laplacian operator

spmeisburger committed:
refactored mdx2.command_line.scale, untested

spmeisburger committed:
implement subsample in mdx2.scale for testing

spmeisburger committed:
function to mask singletons (reflections with only one observation) in scaling

spmeisburger committed:
mask singletons when loading data for scalingg (untested)

spmeisburger committed:
add prescaling refinement to mdx2.scale

spmeisburger committed:
fixed incorrect default value of absorption.enable in scaling

spmeisburger committed:
implement isotropic averaging for prescaling optimization

spmeisburger committed:
added HKLGrid class for file grid accumulation

spmeisburger committed:
added mdx2.reintegrate (untested)

spmeisburger committed:
removed split parameter temporarily

spmeisburger committed:
balance load between parallel jobs and main process

spmeisburger committed:
pre-allocate array to remove accumulation bottleneck

spmeisburger committed:
fixed runtime bug

spmeisburger committed:
force ndiv to be tuple

spmeisburger committed:
add option to output intensity table instead of counts

spmeisburger committed:
add comments

spmeisburger committed:
estimate chunks automatically when creating data.ImageSeries from expt file

spmeisburger committed:
bug fix

spmeisburger committed:
add slab iterator for creating virtual datasets

spmeisburger committed:
adjust chunk shape for final slice when array dimension is not a multiple of the chunk shape

spmeisburger committed:
patch NXvirtualfields to enable reading virtual datasets with any layout

spmeisburger committed:
modify mdx2.tree to use utils.nxload instead of nexus.nxload

spmeisburger committed:
implemented ImageSeries.save method with option to generate virtual dataset and source files

spmeisburger committed:
minor edits, debugging needed still

spmeisburger committed:
print statement for debugging

spmeisburger committed:
more print statement for debugging

spmeisburger committed:
more print statement for debugging

spmeisburger committed:
bug fix

spmeisburger committed:
fix bug in ImageSeries.load

spmeisburger committed:
determine chunks from virtual dataset source file

spmeisburger committed:
in ImageSeries.save, ensure virtual dataset is read back

spmeisburger committed:
add test for saving ImageSeries as virtual dataset

spmeisburger committed:
add source_directory option for saving source files

spmeisburger committed:
implement saving virtual data in mdx2.import_data with datastore option

spmeisburger committed:
fix simple-parsing bug in mdx2.import_data

spmeisburger committed:
add logging feature, implement for mdx2.import_data

spmeisburger committed:
tweaks to logging

spmeisburger committed:
more tweaks to logging

spmeisburger committed:
add option to override default choice for certain chunk dimensions

spmeisburger committed:
attempt to implement geometry for non-scan-varying crystal models

spmeisburger committed:
fix auto-chunk determination for Eiger detectors

spmeisburger committed:
add logging to files for remaining command-line tools

spmeisburger committed:
add ignored sandbox folder

spmeisburger committed:
removed pandas dependency from mdx2.map

spmeisburger committed:
in mdx2.import_data, --datastore datastore is the default instead of None

spmeisburger committed:
in mdx2.import_data, removed legacy multiprocessing-based data import

spmeisburger committed:
removed legacy image reader (read_all, read_all_parallel) from dxtbx_machinery

spmeisburger committed:
add info for version 1.0.3

spmeisburger committed:
removed legacy behavior of mdx2.find_peaks when nproc=1, joblib is always used to execute

spmeisburger committed:
removed legacy behavior of mdx2.mask_peaks when nproc=1, joblib is always used

spmeisburger committed:
removed legacy behavior of mdx2.integrate when nproc=1, now joblib is used exclusively

spmeisburger committed:
updated insulin tutorial readme

Commit:
updated instulin tutorial notebooks

spmeisburger committed:
updated install instructions to use diff-use rather than ando-lab path

spmeisburger committed:
added note to README.md indicating that it is a fork

spmeisburger committed:
remove legacy behavior of mdx2.bin_image_series when nproc=1, now joblib always runs

spmeisburger committed:
streamlined README.md

Commit:
updated multi-crystal example

bielj commented on examples/insulin-tutorial/5-Integration_scaling_and_merging.ipynb:
```diff
@@ -1583,7 +122,7 @@
       "  importing as GridData from mdx2.geometry\n",
       "subtracting background from count rate\n",
-      "correcting solid angle for attenuation\n",
+      "correctin solid angle for attenuation\n",
```
Typo in output here suggests typo introduced in logging during mdx2.correct

bielj commented:
@coderabbitai review

jlee733 commented:
Other than the coderabbit spell check and missing defaults for function parameters, the code looks good @spmeisburger

bielj commented on mdx2/dxtbx_machinery.py:
```diff
@@ -45,6 +47,14 @@ def num_images(self):
         return nimgs
 
+    @property
+    def panel_offset(self):
+        """offset in pixels (stride) of successive panels in the image file"""
+        d = _get_pad_module_gap((self._panel,))
+        row_size = d.module_size_slow + d.gap_slow
+        col_size = d.module_size_fast + d.gap_fast
+        return row_size, col_size
```
This should be created as a separate issue.

spmeisburger replied:
I added compatibility checks in the __init__ function (local commit).

bielj commented on mdx2/utils.py:
```diff
@@ -2,23 +2,35 @@
 # FUNCTIONS FOR LOADING AND SAVING MDX2 CLASSES TO NEXUS FILES
 
 
+def _patch_virtualfields(g):
```
From local CodeRabbit:
Add error handling for file operations and missing paths.

The function lacks error handling for several failure scenarios:

- g.nxfile context manager could fail
- f.get(g.nxpath) could return None if the path doesn't exist
- File read operations could fail

This could cause nxload to fail unexpectedly when encountering malformed files or inaccessible virtual fields.

Apply this diff to add error handling:
```
 def _patch_virtualfields(g):
     """Recursively patch NXvirtualfields in a Nexus group to set their shape."""
     if isinstance(g, NXgroup):
         for entry in g.entries.values():
             _patch_virtualfields(entry)
     elif isinstance(g, NXvirtualfield):
         logger.info(f"patching virtual field: {g.nxpath}")
```

bielj commented on mdx2/command_line/import_data.py:
```diff
@@ -2,51 +2,86 @@
+
+    configure_logging(filename="mdx2.import_data.log")
+    params = parse_arguments(args=args)
+    logging.info("running mdx2.import_data with parameters: %s", params)
```
From local CodeRabbit:
Inconsistent logger usage: Use module logger instead of root logger.

Lines 82 and 84 use logging.info (root logger) instead of logger.info (module logger defined at line 17). This is inconsistent with line 57 which correctly uses logger.info, and will cause these messages to bypass the module-specific logger configuration.

Apply this diff:
```
 def run(args=None):
     """Run the import data script"""

     configure_logging(filename="mdx2.import_data.log")
     params = parse_arguments(args=args)
-    logging.info("running mdx2.import_data with parameters: %s", params)
+    logger.info("running mdx2.import_data with parameters: %s", params)
     run_import_data(params)
-    logging.info("done!")
+    logger.info("done!")
```

bielj commented on mdx2/command_line/import_data.py:
```diff
@@ -2,51 +2,86 @@
+    )
+
+    def write_stack(istart, istop, filename, datapath):
+        source = nxload(filename, "r+")[datapath]
```
From local CodeRabbit:

Resource leak: File handle not closed in parallel execution.

The nxload call opens a file with "r+" mode but never closes it. In parallel execution, this will create multiple open file handles that are never explicitly closed, potentially causing resource exhaustion.

Apply this diff to use a context manager:
```
 def write_stack(istart, istop, filename, datapath):
-    source = nxload(filename, "r+")[datapath]
-    data = iset.read_stack(istart, istop)
-    source[:, :, :] = data
+    with nxload(filename, "r+") as f:
+        source = f[datapath]
+        data = iset.read_stack(istart, istop)
+        source[:, :, :] = data
```

bielj commented on mdx2/command_line/merge.py:
```diff
@@ -94,26 +109,26 @@ def run(args=None):
-    if args.scale is not None:
-        for fn, refiner in zip(args.scale, MR._batch_refiners):
+    if scale is not None:
+        for fn, refiner in zip(scale, MR._batch_refiners):
```
From local CodeRabbit:
Potential mismatch between scale files and batch refiners.

The zip(scale, MR._batch_refiners) silently truncates if the number of scale files doesn't match the number of batches. This could lead to subtle bugs where some batches don't get their scaling models applied.

Consider adding a validation check:
```
     if scale is not None:
+        if len(scale) != len(MR._batch_refiners):
+            raise ValueError(
+                f"Number of scale files ({len(scale)}) does not match "
+                f"number of batches ({len(MR._batch_refiners)})"
+            )
         for fn, refiner in zip(scale, MR._batch_refiners):
```

bielj commented on mdx2/command_line/reintegrate.py:
```diff
@@ -0,0 +1,160 @@
+        tab = tab.bin(count_name="pixels")
+        tab.phi /= tab.pixels
+        tab.iy /= tab.pixels
+        tab.ix /= tab.pixels
```
From local CodeRabbit:
Potential division by zero after binning.

After binning on line 101, some bins could have pixels=0, which would cause a division by zero error when computing averages for phi, iy, and ix. While line 98-99 handles the case where the entire table is empty, individual bins within a non-empty table could still have zero pixels.

Consider adding a safeguard:
```
 tab.phi /= tab.pixels
 tab.iy /= tab.pixels
 tab.ix /= tab.pixels
```
Replace with:
```
+mask = tab.pixels > 0
+tab.phi = np.where(mask, tab.phi / tab.pixels, 0)
+tab.iy = np.where(mask, tab.iy / tab.pixels, 0)
+tab.ix = np.where(mask, tab.ix / tab.pixels, 0)
```
Or filter out zero-pixel bins before division.

bielj commented on mdx2/command_line/reintegrate.py:
```diff
@@ -0,0 +1,160 @@
+    hkl_table.k = hkl_table.k.astype(np.float32)
+    hkl_table.l = hkl_table.l.astype(np.float32)
+    if params.output == "intensity":
+        hkl_table.intensity = (hkl_table.counts - hkl_table.background_counts) / hkl_table.scale
```
From local CodeRabbit:
CodeRabbit
Division by zero or near-zero scale values.

When computing intensity and intensity_error, division by hkl_table.scale can produce inf or nan values if scale is zero or very small. This could happen for reflections with no or minimal scaling corrections.

Add validation before division:
```
 if params.output == "intensity":
+    valid = hkl_table.scale > 0
+    if not np.all(valid):
+        logger.warning(f"Found {(~valid).sum()} reflections with zero or negative scale, filtering them out")
+        hkl_table = hkl_table[valid]
     hkl_table.intensity = (hkl_table.counts - hkl_table.background_counts) / hkl_table.scale
     hkl_table.intensity_error = np.sqrt(hkl_table.counts) / hkl_table.scale
```

bielj commented on mdx2/command_line/scale.py:
```diff
@@ -2,328 +2,288 @@
+    if opts.parameters.outfile is None:
+        opts.parameters.outfile = generate_default_outfiles(opts.parameters.hkl)
+        if opts.parameters.outfile is None:
+            raise SystemExit("unable to auto-generate output file names from input name pattern")
```
From local CodeRabbit:
Handle empty list case for outfile.

With nargs="*" on line 99, if no output files are provided, opts.parameters.outfile may be an empty list [] rather than None. The condition should check for both cases.
```
-    if opts.parameters.outfile is None:
+    if not opts.parameters.outfile:
         opts.parameters.outfile = generate_default_outfiles(opts.parameters.hkl)
         if opts.parameters.outfile is None:
             raise SystemExit("unable to auto-generate output file names from input name pattern")
```

bielj commented on mdx2/command_line/scale.py:
```diff
@@ -2,328 +2,288 @@
+        index_map = np.floor(hkl.s * 1000).astype(np.uint16)
+    else:
+        print("Grouping redundant observations")
+        _, index_map, _ = hkl.unique()
```
From local CodeRabbit:
Potential integer overflow in isotropic binning.

Line 283 uses np.uint16 which has a max value of 65535. If hkl.s * 1000 exceeds this (i.e., s > 65.535), there will be silent overflow causing incorrect binning. Consider using np.uint32 or adding a bounds check.

Also, the hardcoded 1000 bin width multiplier (as noted in the TODO) should eventually be configurable.
```
-        index_map = np.floor(hkl.s * 1000).astype(np.uint16)
+        index_map = np.floor(hkl.s * 1000).astype(np.uint32)
```

bielj commented on mdx2/command_line/__init__.py:
```diff
@@ -0,0 +1,25 @@
+    handlers.append(console_handler)
+
+    # Configure root logger
+    logging.basicConfig(level=level, handlers=handlers)
```
From local CodeRabbit:
Potential silent failure if logging is already configured.

logging.basicConfig() only takes effect if the root logger has no handlers. If this function is called multiple times or after other logging configuration, subsequent calls will silently do nothing, potentially causing confusing behavior.

Does Python logging.basicConfig have a force parameter to reconfigure logging?
If your codebase supports Python 3.8+, consider adding force=True:
```
-logging.basicConfig(level=level, handlers=handlers)
+logging.basicConfig(level=level, handlers=handlers, force=True)
```
Alternatively, manually clear existing handlers before configuration:
```
root_logger = logging.getLogger()
for handler in root_logger.handlers[:]:
    root_logger.removeHandler(handler)
```

bielj commented on tests/test_nexus_files.py:
```diff
@@ -17,3 +19,28 @@ def test_writing_data_after_nxsave(tmp_path):
+    # write some data to the second source file, check that it appears in the virtual dataset
+    source_series = ImageSeries.load(nxobj.data._vfiles[1], mode="r+")
+    source_series.data[0, 0, 0] = 666
+    assert image_series.data[8, 0, 0] == 666
```
From local CodeRabbit:
Close the file handle to prevent resource leaks.

The code opens a source file at line 44 but never closes it, which can cause resource leaks or test failures (especially on Windows where open file handles can prevent cleanup). Additionally, line 44 uses the private attribute _vfiles.

Apply this diff to use a context manager and ensure proper cleanup:
```
-    # write some data to the second source file, check that it appears in the virtual dataset
-    source_series = ImageSeries.load(nxobj.data._vfiles[1], mode="r+")
-    source_series.data[0, 0, 0] = 666
-    assert image_series.data[8, 0, 0] == 666
+    # write some data to the second source file, check that it appears in the virtual dataset
+    with ImageSeries.load(nxobj.data._vfiles[1], mode="r+") as source_series:
+        source_series.data[0, 0, 0] = 666
+    assert image_series.data[8, 0, 0] == 666
```

bielj commented:
Added some missing prompts from my local run of CodeRabbit, which I thought were worth pointing out or noticing.
I've looked through everything and have finished my initial review.

spmeisburger committed:
offer clearer explanation for why sys.path.remove is necessary

spmeisburger commented on examples/insulin-multi-crystal/scale_factor_mdx2_vs_dials.ipynb:
```diff
@@ -13,49 +13,11 @@
+   "outputs": [],
    "source": [
+    "import sys\n",
+    "sys.path.remove('') # needed for some reason.. this shadows the mdx2 import\n",
```
I made the changes in all occurrences: 
- examples/insulin-multi-crystal/cchalf_vs_redundancy.ipynb
- examples/insulin-multi-crystal/map_statistics.ipynb
- examples/insulin-multi-crystal/scale_factor_mdx2_vs_dials.ipynb
- examples/insulin-multi-crystal/scaling_model.ipynb
- examples/insulin-multi-crystal/visualization.ipynb

spmeisburger committed:
Optional fields missing default values

spmeisburger commented on mdx2/command_line/bin_image_series.py:
```diff
@@ -2,50 +2,72 @@
+    data: str = field(positional=True)  # NeXus data file containing the image_series
+    bins: Tuple[int, int, int] = field(positional=True)  # number per bin in each direction (frames, y, x)
+    mask: Optional[str]  # name of NeXus file containing the mask
+    valid_range: Optional[Tuple[int, int]]  # minimum and maximum valid data values
```
Changes committed to my local copy, however we lack tests for this particular use case (+/- mask, +/- valid_range flags). TO DO: create an issue describing situations that need to be tested.

spmeisburger committed:
Optional fields missing default values

spmeisburger commented on mdx2/command_line/correct.py:
```diff
@@ -2,63 +2,69 @@
 
+    geom: str = field(positional=True)  # NeXus data file containing miller_index
+    hkl: str = field(positional=True)  # NeXus data file containing hkl_table
+    background: Optional[str]  # NeXus file with background map
```
Changes committed to my local copy. Again, we lack tests for this particular use case (+/- background). Add this to the list of things to be tested.

spmeisburger committed:
fixed typo

spmeisburger committed:
fixed typo

spmeisburger committed:
Fixed missing default value for optional mask field.

spmeisburger commented on mdx2/command_line/integrate.py:
```diff
@@ -2,55 +2,63 @@
-    args = parser.parse_args(args)
+    geom: str = field(positional=True)  # NeXus data file containing miller_index
+    data: str = field(positional=True)  # NeXus data file containing image_series
+    mask: Optional[str]  # NeXus data file containing mask
```
Fixed, but again we do not have tests for this use case.

spmeisburger committed:
log a warning if division by zero when computing intensity

spmeisburger commented on mdx2/command_line/reintegrate.py:
```diff
@@ -0,0 +1,160 @@
+            grid.accumulate_from_table(tab, resize=False)
+
+    print(f"Saving table of integrated data to {params.outfile}")
+    hkl_table = grid.to_table(sparse=True)
+    hkl_table.h = hkl_table.h.astype(np.float32)
+    hkl_table.k = hkl_table.k.astype(np.float32)
+    hkl_table.l = hkl_table.l.astype(np.float32)
+    if params.output == "intensity":
+        hkl_table.intensity = (hkl_table.counts - hkl_table.background_counts) / hkl_table.scale
+        hkl_table.intensity_error = np.sqrt(hkl_table.counts) / hkl_table.scale
+        del hkl_table.counts, hkl_table.background_counts, hkl_table.scale
```
I think it's pretty unlikely that the scale column would contain a zero, but it could happen (and would signal a likely bug elsewhere in the code). This is a reasonable solution, I committed it to my local copy.

spmeisburger committed:
fixed docstring typo

spmeisburger committed:
organized imports

spmeisburger committed:
added detailed detector compatibility-checks to Experiment wrapper

spmeisburger committed:
fixed URL link to diffuse.science

spmeisburger commented on README.md:
```diff
@@ -1,27 +1,36 @@
+<!--[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.10519719.svg)](https://doi.org/10.5281/zenodo.10519719)-->
+
+> [!NOTE]
+> This is a fork of [ando-lab/mdx2](https://github.com/ando-lab/mdx2), intended for active development by [The Diffuse Project](https://github.com/ando-lab/mdx2) 
```
Fixed Diffuse Project link so it points to https://diffuse.science

spmeisburger committed:
fixed typo

spmeisburger commented on README.md:
```diff
@@ -1,27 +1,36 @@
+- Performance boost for `mdx2.import_data` using parallel read and write. The `data.nxs` file contains a virtual dataset linking to neXus files in a subdirectory (`datastore/` by default).
+- `mdx2.reintegrate` -- New command-line tool to create fine maps after scaling (single-sweep only: multi-crystal datasets not yet implemented)
+- Optional pre-scaling in `mdx2.scale` to correct anisotropic background
+- Improved handling of command-line arguments via `dataclass` attributes and `simpleparsing` package
```
Fixed

spmeisburger committed:
switched logger to loguru

spmeisburger committed:
cleaned up logging output

spmeisburger committed:
centralized logging in command-line routines

spmeisburger committed:
moved all logging and parallel processing logic to command-line tools

spmeisburger committed:
created more informative logging mesages

spmeisburger committed:
cache results of _get_pad_module_gap()

spmeisburger committed:
improve comment clarity about loadobj vs nxload usage

spmeisburger committed:
rename variables to avoid l as a variable name

spmeisburger committed:
improve comment clarity about loadobj vs nxload usage

spmeisburger committed:
remove old commented-out code

spmeisburger committed:
avoid in‑place mutation of Cinterp

spmeisburger committed:
use typing.Tuple instead of tuple for consistency across codebase

spmeisburger committed:
Fix AttributeError: access binned.data.shape instead of binned.shape

spmeisburger committed:
replaced all private parallel._backend attribute accesses with the public joblib.parallel.get_active_backend() API across all 6 command-line files that use parallel processing

spmeisburger committed:
raise informative errors if empty table lists are encountered

spmeisburger committed:
add default value for optional arguments

spmeisburger committed:
added default value of None for split option

spmeisburger committed:
replace self.data.__deepcopy__() with deepcopy(self.data)

spmeisburger committed:
switched to public dxtbx function to get panel gaps

spmeisburger committed:
properly handle out of bounds case when interpolating over phi values

spmeisburger commented on mdx2/dxtbx_machinery.py:
```diff
@@ -270,10 +331,8 @@ def frame_shape(self):
         return (ny, nx)
 
-    def read_frame(self, ind, verbose=True, maskval=-1):
+    def read_frame(self, ind, maskval=-1):
```
Based on my search results:

- No breaking changes found: There are no call sites in the codebase that pass a verbose parameter to read_frame
- Single internal usage: The only call to read_frame is within the same class (ImageSet.read_stack at line 350), and it doesn't use verbose
- No external usage: No other files in the codebase call read_frame method

The current signature def read_frame(self, ind, maskval=-1): appears to be correct and there are no call sites that need updating. The verbose parameter was likely removed in a previous refactoring, and all code has already been adapted.

Conclusion: No changes are needed. The codebase is already consistent with the current read_frame signature without the verbose parameter.

As a note: this part of the API is only used internally by mdx2, and since we are still in active development, we don't need to worry about DeprecationWarnings for this part of the package.

spmeisburger commented on mdx2/dxtbx_machinery.py:
```diff
@@ -1,11 +1,10 @@
 import numpy as np
 from cctbx.eltbx import attenuation_coefficient
 from dxtbx import flumpy
+from dxtbx.format.FormatPilatusHelpers import sensor_active_areas
```
This is part of a larger issue that we don't pin versions for most of our dependencies. We have not decided on a good solution for this yet, it needs further discussion / work and is beyond the scope of this PR.

spmeisburger committed:
fix HKLTable.concatenate: correct data_keys intersection and validate ndiv compatibility

spmeisburger committed:
fixed formatting error in loguru log message

spmeisburger committed:
add validation for positive bin sizes in bin_image_series with tests

spmeisburger commented on mdx2/command_line/bin_image_series.py:
```diff
@@ -2,50 +2,124 @@
-    nxs = saveobj(binned, args.outfile, name="binned_image_series")
+    logger.info("Binning images with bin size: {}", bins)
+
+    # Compute binning parameters
+    bins_array = np.array(bins)
+    nbins = np.ceil(image_series.shape / bins_array).astype(int)
+    sl_0 = slice_sections(image_series.shape[0], nbins[0])
+    sl_1 = slice_sections(image_series.shape[1], nbins[1])
+    sl_2 = slice_sections(image_series.shape[2], nbins[2])
+
```
I fixed this issue by adding a `__post_init__(self)` function to the Parameters dataclass, which is superior because the validation checks are performed outside of parse_arguments. This pattern might be useful in other instances: we could move the validation and parameter interdependence logic to the dataclass itself.

spmeisburger committed:
improve type annotations for list-valued fields in merge Parameters

spmeisburger committed:
use context manager for HDF5 file access in chunks property

spmeisburger committed:
reuse slices list instead of recreating iterator in mask_peaks

spmeisburger committed:
prefix unused SVD component with underscore in find_peaks

spmeisburger committed:
Fix joblib backend logging and centralize implementation

- Add log_parallel_backend() helper to command_line/__init__.py
- Update all command-line tools to use centralized logging
- Remove incorrect use of get_active_backend() which returned stale values

spmeisburger committed:
Fix masked scalar detection in bin_image_series

Replace isinstance check with np.ma.is_masked() to properly handle
fully-masked slices that return np.ma.masked singleton

spmeisburger committed:
Configure default logging for utils module

Set WARNING level and simplified format for non-CLI usage

spmeisburger committed:
Refactor: Move I/O functions from utils to new io module

- Create new mdx2/io.py with file I/O functions (nxload, nxsave, loadobj, saveobj, _patch_virtualfields)
- Keep numerical/computational functions in mdx2/utils.py (interpolation, slice_sections)
- Update all imports across 12 command-line tools, core modules, tests, and example notebooks
- All tests pass successfully (34/34)

spmeisburger committed:
Add backward compatibility for I/O imports from utils

- Re-export I/O functions (nxload, nxsave, loadobj, saveobj) from mdx2.utils
- Use __getattr__ for lazy loading with deprecation warnings
- Users can still import from mdx2.utils but get clear migration guidance
- Maintains compatibility with existing user notebooks and scripts
- All tests pass (34/34)

spmeisburger committed:
Fix code quality issues: type annotations, docstring, unused vars, and code style

spmeisburger committed:
Refactor scale.py: move parameter validation to __post_init__

Move mca2020 shortcut logic and outfile auto-generation from
parse_arguments into Parameters.__post_init__. This encapsulates
inter-parameter validation within the dataclass, improving reusability
and testability when Parameters is instantiated directly.

spmeisburger committed:
Refactor merge.py and simplify scale.py: move validation to __post_init__

- Add __post_init__ to merge.py Parameters to validate Friedel split
  requires geometry parameter (encapsulates inter-parameter dependency)
- Simplify parse_arguments in both scale.py and merge.py (remove
  exception handling as simple-parsing handles ValueError automatically)
- Update test to expect ValueError instead of SystemExit for validation
  errors raised in __post_init__

spmeisburger committed:
Add public batch_refiners API to BatchModelRefiner

- Add batch_refiners property to BatchModelRefiner class that returns
  an iterator over individual refiners
- Update scale.py to use batch_refiners instead of _batch_refiners
  (2 locations: model transfer and saving)
- Update merge.py to use batch_refiners for loading scaling models
- Improves encapsulation by removing direct access to private
  _batch_refiners attribute throughout the codebase

spmeisburger committed:
removed unused return value

spmeisburger committed:
add note to docstring warning that the save method mutates self.data

spmeisburger committed:
Add sample_spacing validation to import_geometry Parameters

- Add __post_init__ method to validate all sample_spacing values are
  positive (prevents downstream geometry computation issues)
- Add tests for negative and zero sample_spacing values
- Follows established pattern of validation in __post_init__ rather
  than parse_arguments for better encapsulation and testability

spmeisburger committed:
Add error handling for missing to_nexus method in saveobj

- Check for to_nexus method existence before calling it
- Raise TypeError with clear message if method is missing
- Improves user experience by providing informative error messages
  instead of generic AttributeError

spmeisburger committed:
Fix reintegrate.py to handle missing scaling models gracefully

Previously, reintegrate.py would crash with a NeXusError when given a
scaling model file that doesn't contain all possible correction models
(e.g., missing absorption_model, offset_model, etc.).

This fix checks for the existence of each model in the scale file before
attempting to load it, following the same pattern used successfully in
merge.py. Models that don't exist in the file are set to None and handled
gracefully during calculations.

Also removed obsolete TODO comment and fixed minor formatting in io.py.

spmeisburger committed:
Refactor: extract calc_corrections to module-level function

- Improves testability by allowing independent unit testing
- Enables reuse across other modules (integrate.py, correct.py)
- Adds comprehensive docstring with full API documentation

spmeisburger committed:
Remove loguru logging from io.py module

This commit addresses issues with loguru logger configuration in the io.py
module that could cause unintended side effects when imported in contexts
like Jupyter notebooks.

Changes to mdx2/io.py:
- Remove loguru import and logger configuration (lines 11-17)
- Replace logger.warning() with Python's warnings.warn() for version
  mismatch notifications
- Remove logger.debug() call in _patch_virtualfields() and replace with
  explanatory comment
- Remove logger.info() calls from loadobj() and saveobj() functions that
  cluttered logs when processing many files

Changes to tests/test_version_number.py:
- Update test_mdx2_utils_nxload() to use pytest.warns() instead of
  capturing loguru output
- Remove loguru import and StringIO dependency
- Add pytest import for warning assertions

Changes to command-line tools:
Add explicit "Saving ... to {file}..." log messages before saveobj()
calls in the following command-line tools to provide user feedback:
- import_geometry.py
- find_peaks.py
- merge.py
- mask_peaks.py
- bin_image_series.py
- correct.py
- map.py
- reintegrate.py

Benefits:
- Eliminates side effects from module-level logger configuration
- Reduces log clutter when loading/saving multiple files
- Uses standard Python warnings that users can control via warnings module
- Maintains backward compatibility with pytest for version mismatch checks
- Improves user feedback in command-line tools

All tests pass (36/36).

spmeisburger committed:
Add missing files from logging refactor

- Add log message to integrate.py for saving integrated data
- Fix trailing comma formatting in io.py warnings.warn() call

spmeisburger committed:
Validate output file count matches input file count in scale command

Add validation in Parameters.__post_init__() to ensure len(outfile) equals
len(hkl), preventing truncation errors when zipping with MR.batch_refiners.

spmeisburger committed:
Improve error handling for virtual dataset access

- Add public properties virtual_source_files and virtual_dataset_path on ImageSeries
- Raise RuntimeError on AttributeError (API changes) instead of silent None
- Raise FileNotFoundError when virtual dataset source files are missing
- Improve error messages to distinguish API changes from file access issues
- Update import_data.py to use new properties instead of accessing internal attributes

spmeisburger committed:
Add test coverage for optional mask/background parameters

- Add test_correct_parse_arguments with 4 test cases (with/without background, various correction flags)
- Add test_reintegrate_parse_arguments with 5 test cases (all combinations of mask/background)
- Addresses review feedback about missing test coverage for optional CLI parameters
- All 46 tests now pass (9 new tests added)

spmeisburger committed:
formatting

spmeisburger committed:
Remove redundant version check from loadobj()

- loadobj() was re-reading and checking mdx2_version after nxload() already did it
- The duplicate check was a no-op (pass with TODO comment)
- Version checking is now centralized in nxload() which issues warnings
- Added comment documenting that version checking is handled by nxload()
- No behavior change: the redundant check did nothing anyway
- Addresses review feedback about duplicate version checking logic

spmeisburger committed:
Remove bold formatting from log messages

spmeisburger committed:
Refactor: reduce command-line parsing boilerplate with factory function

Add make_argument_parser() factory to generate standardized parse_arguments()
functions across all command-line tools. Reduces boilerplate from 6 lines to 1
line per module while maintaining backward compatibility and test coverage.

- Applied uniform ArgumentParser configuration (supports nested dataclasses)
- All 41 tests pass without modification
- Cleaner, more maintainable code following DRY principle

spmeisburger committed:
Add parameter validation to command-line tools

Add __post_init__ validation for numeric parameters in 6 command-line tools:
- bin_image_series: valid_range min < max
- find_peaks: sigma_cutoff > 0
- mask_peaks: sigma_cutoff > 0
- integrate: subdivide elements > 0, max_spread > 0
- map: limits tuple validation (min < max for h, k, l)
- merge: outlier > 0 when provided
- reintegrate: subdivide elements > 0

Includes 16 new test cases verifying validation behavior.

spmeisburger committed:
applied linting

spmeisburger committed:
Security: Add module validation to loadobj function

Prevent arbitrary code execution vulnerability by:
- Restricting imports to mdx2.* modules only
- Validating that loaded classes have a callable from_nexus method
- Adding comprehensive docstring with parameter and exception documentation

Includes test case verifying rejection of untrusted module names.

spmeisburger committed:
Add explicit duplicate file validation to scale command

Improve error handling for duplicate input files by:
- Adding explicit duplicate check in __post_init__ that runs before output file generation
- Providing clear error message: 'Duplicate input files are not allowed'
- Preserving valid use case: same filename in different directories
- Adding inline comments documenting edge case behavior

Includes test coverage for duplicate file scenarios.

spmeisburger committed:
Add comment explaining NXfield pickling behavior in reintegrate

NXfield objects from nexusformat can be safely pickled and used with
joblib.Parallel (nproc > 1) because they implement custom __getstate__
and __setstate__ methods that serialize only metadata (filename, path,
shape, dtype) rather than the actual HDF5 file handle. Each worker
process reopens the file lazily when data is accessed.

This clarifies why passing the mask NXfield to parallel workers works
correctly, addressing a question raised during code review.

See: https://github.com/nexpy/nexusformat/blob/main/src/nexusformat/nexus/tree.py#L2100-L2125

spmeisburger committed:
Handle no peaks found case with clear error message

Previously, if no peaks were found above the count_threshold,
Peaks.stack([]) would raise a cryptic ValueError from np.concatenate.

Now we check for this case explicitly and raise a helpful error message
directing users to adjust their threshold or check their data.

This prevents downstream failures in mask_peaks.py which expects valid
peak_model and peaks objects in the output file. Since fitting a Gaussian
peak model requires at least some data points, treating this as an error
rather than silently succeeding is the appropriate behavior.

spmeisburger committed:
Fix regex pattern and add test for missing from_nexus method

- Use raw string for regex pattern to properly escape metacharacters
  in test_loadobj_security_checks (addresses RUF043)
- Add Test 3 to verify TypeError is raised when a class exists but
  lacks a callable from_nexus method, exercising the security
  validation path in loadobj()
- Create a test module fixture with ClassWithoutFromNexus to test
  this edge case without modifying production code

spmeisburger committed:
linting

spmeisburger committed:
added dev group to optional dependencies in pyproject.toml that includes ruff and pytest, updated README.md developer install instructions

spmeisburger committed:
remove redundant definitions of model_name in both branches of if, else

spmeisburger committed:
Improve help text for nproc parameter to cover negative nubmers (n_jobs in joblib Parallel)

spmeisburger commented on mdx2/scaling.py:
```diff
@@ -129,7 +129,31 @@ def Bz(self):
 
     @property
     def B(self):
-        raise NotImplementedError
+        Nx, Ny, Nz = self.shape
+        k, l, m = np.meshgrid(np.arange(Nx), np.arange(Ny), np.arange(Nz), indexing="ij")
+
+        # Neighbor indices
+        right = self._sub2ind(k[1:, :, :], l[1:, :, :], m[1:, :, :]).ravel(order="F")
+        left = self._sub2ind(k[:-1, :, :], l[:-1, :, :], m[:-1, :, :]).ravel(order="F")
+        up = self._sub2ind(k[:, 1:, :], l[:, 1:, :], m[:, 1:, :]).ravel(order="F")
+        down = self._sub2ind(k[:, :-1, :], l[:, :-1, :], m[:, :-1, :]).ravel(order="F")
+        front = self._sub2ind(k[:, :, :-1], l[:, :, :-1], m[:, :, :-1]).ravel(order="F")
+        back = self._sub2ind(k[:, :, 1:], l[:, :, 1:], m[:, :, 1:]).ravel(order="F")
+
+        # Flatten all neighbor indices
+        row_index = np.concatenate((left, right, down, up, front, back))
+        col_index = np.concatenate((right, left, up, down, back, front))
+        vals = np.ones_like(row_index, dtype=np.double)
+        shape = (Nx * Ny * Nz, Nx * Ny * Nz)
+        L = sparse.coo_matrix((vals, (row_index, col_index)), shape=shape)
+
+        # Weight according to number of neighbors
+        nn = np.asarray(L.sum(axis=1)).flatten()
+        W = sparse.diags(1 / nn)
+        L = W @ L
+
+        # Center minus average of neighbors
+        return sparse.eye(shape[0]) - L
```
The edge case of a 1x1x1 grid seems very unlikely, and not worth addressing here. I'll add a comment in the code that Nx=Ny=Nz may lead to unexpected behavior, in case it comes up again in future.

spmeisburger committed:
add note in code about laplacian matrix construction failing for 1x1x1 edge case

spmeisburger commented on mdx2/command_line/bin_image_series.py:
```diff
@@ -2,50 +2,118 @@
+    new_iy = np.array([image_series.iy[sl].mean() for sl in sl_1])
+    new_ix = np.array([image_series.ix[sl].mean() for sl in sl_2])
+
+    def binslab(sl):
+        outslab = np.empty([len(sl_1), len(sl_2)], dtype=float)
+        tmp = image_series._as_np(image_series.data[sl, :, :])
+        tmp = np.ma.masked_equal(tmp, image_series._maskval, copy=False)
+        if valid_range is not None:
+            tmp = np.ma.masked_outside(tmp, valid_range[0], valid_range[1], copy=False)
+        if mask is not None:
+            msk = image_series._as_np(mask[sl, :, :])
+            tmp = np.ma.masked_where(msk, tmp, copy=False)
+        for ind_y, sl_y in enumerate(sl_1):  # not vectorized - could be slow?
+            for ind_x, sl_x in enumerate(sl_2):
+                val = tmp[:, sl_y, sl_x].mean()
+                if np.ma.is_masked(val):
+                    val = np.nan
+                outslab[ind_y, ind_x] = val
+        return outslab
```
Let's defer addressing this until later. The code referenced is likely to change soon. 

Background: This cross-module dependency on private methods appeared because the parallel logic (i.e. joblib) was refactored. It originally appeared in the main data module (mdx2.data), where access to private functions was not an issue. It was subsequently moved to the command_line tools. 

What's likely to change: We ought to refactor the code in binslab. The numerical parts should be moved to a numpy-only function for binning (one that accepts and returns numpy arrays), and the remaining dependencies on private functions should be addressed after refactoring. The advantage is that binslab could then be compiled using numba and perhaps reused elsewhere. I'll open an issue to address this.

spmeisburger commented on mdx2/command_line/reintegrate.py:
```diff
@@ -0,0 +1,255 @@
+        absorption_model = None
+        offset_model = None
+        detector_model = None
+    mask = nxload(params.mask).entry.mask.signal if params.mask else None
+
+    def intchunk(sl):
+        ims = image_series[sl]
+        if mask is not None:
+            tab = ims.index(miller_index, mask=mask[sl].nxdata)  # added nxdata to deal with NXfield wrapper
+        else:
+            tab = ims.index(miller_index)
+        if len(tab) == 0:
+            return None  # signal that no pixels were integrated
+        tab.ndiv = params.subdivide
+        tab = tab.bin(count_name="pixels")
+        tab.phi /= tab.pixels
+        tab.iy /= tab.pixels
+        tab.ix /= tab.pixels
+        tab = calc_corrections(
+            tab,
+            crystal,
+            corrections,
+            symmetry,
+            scaling_model,
+            absorption_model,
+            detector_model,
+            offset_model,
+            background,
+        )
```
Actually, this line is not a problem -- the code runs fine with nproc > 1. I was initially confused about this, but then discovered that the mask object (NXField) implements custom pickling. The following comments were added to the code, in case this question comes up again:

```python
    # NXfield objects can be pickled safely (they serialize metadata and reopen the file in each worker)
    # so this works correctly with nproc > 1 despite being passed to joblib.Parallel
    mask = nxload(params.mask).entry.mask.signal if params.mask else None
```

spmeisburger commented on mdx2/command_line/reintegrate.py:
```diff
@@ -0,0 +1,255 @@
+
+@dataclass
+class Parameters:
+    """Options for integrating counts on a Miller index grid"""
+
+    geom: str = field(positional=True)  # NeXus data file containing miller_index
+    data: str = field(positional=True)  # NeXus data file containing image_series
+    mask: Optional[str] = None  # NeXus data file containing mask
+    subdivide: Tuple[int, int, int] = (1, 1, 1)  # subdivisions of the Miller index grid
+    scale: Optional[str] = None  # NeXus file with scaling model
+    background: Optional[str] = None  # NeXus file with background binned_image_series
+    nproc: int = 1  # number of parallel processes
+    output: Literal["counts", "intensity"] = "counts"  # counts or intensity
+    outfile: str = "reintegrated.nxs"  # name of the output data
+
+
+def parse_arguments(args=None):
+    """Parse commandline arguments"""
+    parser = ArgumentParser(description=__doc__)
+    parser.add_arguments(Parameters, dest="parameters")
+    opts = parser.parse_args(args)
+    return opts.parameters
+
```
The subdivide parameter is now properly validated. The following code was inserted:
    
```python
    def __post_init__(self):
        """Validate subdivide parameter"""
        for i, div in enumerate(self.subdivide):
            if div <= 0:
                raise ValueError(f"subdivide[{i}] must be > 0, got {div}")
 ```
 
 The `nproc` parameter is actually not limited to positive integers. I updated the help text here (and in all other dataclasses with an nproc parameter) to explain this:
 
```python
     nproc: int = 1  # number of parallel processes (or 1 for sequential, -1 for all CPUs, -N for all but N+1)
```

spmeisburger commented on mdx2/command_line/find_peaks.py:
```diff
@@ -2,72 +2,108 @@
+    """Parse commandline arguments"""
+    parser = ArgumentParser(description=__doc__)
+    parser.add_arguments(Parameters, dest="parameters")
+    opts = parser.parse_args(args)
+    return opts.parameters
+
+
+def run_find_peaks(params):
+    """Run the find peaks script"""
+    geom = params.geom
+    data = params.data
+    count_threshold = params.count_threshold
+    sigma_cutoff = params.sigma_cutoff
+    outfile = params.outfile
+    nproc = params.nproc
+
+    logger.info("Loading geometry and image data...")
+    MI = loadobj(geom, "miller_index")
+    IS = loadobj(data, "image_series")
+
+    logger.info("Finding pixels above threshold: {}", count_threshold)
+
+    # Find peaks in parallel
+    def peaksearch(sl):
+        ims = IS[sl]
+        im_data = ims.data_masked
+        peaks = Peaks.where(im_data > count_threshold, ims.phi, ims.iy, ims.ix)
+        if peaks.size:
+            return peaks
+
+    slices = list(IS.chunk_slice_iterator())
+    logger.info("Searching for peaks in {} image chunks (requested n_jobs: {})...", len(slices), nproc)
+    with Parallel(n_jobs=nproc, verbose=10) as parallel:
+        log_parallel_backend(parallel)
+        peaklist = parallel(delayed(peaksearch)(sl) for sl in slices)
+    logger.info("Peak search completed")
+
+    P = Peaks.stack([p for p in peaklist if p is not None])
+    logger.info("Found {} peak pixels", P.size)
+
```
This was fixed with commit 5c1483ca732c1b96f7b40171755b095f54344d84

spmeisburger commented:
It looks like the major issues have been addressed already, but I still need to re-run the examples with the latest version of the code. This will take some time. Once finished, I'll push the changes and squash / merge.

spmeisburger committed:
allow tight limits in mdx2.mask, where Miller index min == max (fixes broken 6-Visualization.pynb in insulin-tutorial)

spmeisburger committed:
print actual number of cpus used in parallel processing when nproc < 0

spmeisburger committed:
skip logging if a command-line tool is called using the --help or -h flags

jlee733 commented on mdx2/command_line/map.py:
```diff
@@ -2,56 +2,60 @@
 
-    # Required arguments
-    parser.add_argument("geom", help="NeXus file with symmetry and crystal")
-    parser.add_argument("hkl", help="NeXus file with hkl_table")
-    parser.add_argument("--symmetry", default="True", metavar="TF", help="apply symmetry operators?")
-    parser.add_argument(
-        "--limits",
-        default=[0, 10, 0, 10, 0, 10],
-        type=float,
-        nargs=6,
-        metavar=("H1", "H2", "K1", "K2", "L1", "L2"),
-        help="region to map",
-    )
-    parser.add_argument("--signal", default="intensity", help="column in hkl_table to map")
-    parser.add_argument("--outfile", default="map.nxs", help="name of the output NeXus file")
+    def __post_init__(self):
+        """Validate limits parameter"""
+        hmin, hmax, kmin, kmax, lmin, lmax = self.limits
+        if hmin > hmax:
+            raise ValueError(f"limits: hmin must be <= hmax, got hmin={hmin}, hmax={hmax}")
+        if kmin > kmax:
+            raise ValueError(f"limits: kmin must be <= kmax, got kmin={kmin}, kmax={kmax}")
+        if lmin > lmax:
+            raise ValueError(f"limits: lmin must be <= lmax, got lmin={lmin}, lmax={lmax}")
 
```
This would be good a good change for future troubleshooting

spmeisburger replied:
In this case, the checks are performed by simple-parsing using the type hints in the `@dataclass`, which enforces tuple of integers of length 6. Here's a quick test to demonstrate the behavior:

```python
from mdx2.command_line.map import parse_arguments
parse_arguments('geometry.nxs stuff.nxs --limits 0 1 0 1 0'.split()) # incorrect number of limits
```

Output:
```
usage:  [-h] [--symmetry bool] [--limits float float float float float float] [--signal str]
        [--outfile str]
        str str
: error: argument --limits: expected 6 arguments
```

As a general strategy for maintainable code, I try to make sure the `__post_init__` checks are not redundant with the type hints. Although, in a use case where the Parameters are assigned outside a command-line interface, we would want to implement strict checking (using a package such as pydantic). But for now, the user interaction is always through parse_arguments, so I think the code is fine as-is.

spmeisburger replied:
Hi @coderabbitai, I consider this comment resolved. I noticed you raised it again after the most recent commit -- can you tag this discussion there, and resolve both conversations?

spmeisburger replied:
Oh, hang on, I was confused, there is no duplicate conversation. I'll go ahead and resolve this one.

spmeisburger committed:
fix --help checking in case where run() is called from the command-line

bielj commented on mdx2/command_line/import_data.py:
```diff
@@ -2,51 +2,95 @@
+        source_directory=datastore,
+    )
+
+    def write_stack(istart, istop, filename, datapath):
+        source = nxload(filename, "r+")[datapath]
+        data = iset.read_stack(istart, istop)
+        source[:, :, :] = data
```
From CodeRabbit:
Resource leak: File handle not closed in parallel execution.
The nxload call opens a file with "r+" mode but never closes it. In parallel execution, this will create multiple open file handles that are never explicitly closed, potentially causing resource exhaustion.
Apply this diff to use a context manager:
```def write_stack(istart, istop, filename, datapath):
-    source = nxload(filename, "r+")[datapath]
-    data = iset.read_stack(istart, istop)
-    source[:, :, :] = data
+    with nxload(filename, "r+") as f:
+        source = f[datapath]
+        data = iset.read_stack(istart, istop)
+        source[:, :, :] = data
```

spmeisburger replied:
The nxload does not leave files open, and it doesn't work as a context manager. Here are the important lines in the nexusformat source code:

https://github.com/nexpy/nexusformat/blob/3ceae32199a2763a04c18e5732d584b5b3550726/src/nexusformat/nexus/tree.py#L8481-L8483

bielj commented on mdx2/command_line/reintegrate.py:
```diff
@@ -0,0 +1,258 @@
+            return None  # signal that no pixels were integrated
+        tab.ndiv = params.subdivide
+        tab = tab.bin(count_name="pixels")
+        tab.phi /= tab.pixels
+        tab.iy /= tab.pixels
+        tab.ix /= tab.pixels
```
From local CodeRabbit:
Potential division by zero after binning.
After binning on line 101, some bins could have pixels=0, which would cause a division by zero error when computing averages for phi, iy, and ix. While line 98-99 handles the case where the entire table is empty, individual bins within a non-empty table could still have zero pixels.
``` tab.phi /= tab.pixels
 tab.iy /= tab.pixels
 tab.ix /= tab.pixels
```
Replace with:
```
+mask = tab.pixels > 0
+tab.phi = np.where(mask, tab.phi / tab.pixels, 0)
+tab.iy = np.where(mask, tab.iy / tab.pixels, 0)
+tab.ix = np.where(mask, tab.ix / tab.pixels, 0)
```

spmeisburger replied:
The full context here is:

```python
        tab = tab.bin(count_name="pixels")
        tab.phi /= tab.pixels
        tab.iy /= tab.pixels
        tab.ix /= tab.pixels
        tab = calc_corrections( 
            tab,
        # < more args >
        )
```

The function `tab.bin` bins the columns by unique Miller indices, and the count is assigned to the column 'pixels'. Thus, pixels will never be zero. 

However, there's a different possible failure mode. If `tab` is an empty HKLTable (which could happen if all the pixels in this chunk are masked out), the `/=` lines are OK (because numpy allows you to divide two empty arrays, returning another empty array). But the subsequent call to `calc_corrections` might be problematic -- I'm not sure. We should probably refactor `calc_corrections` and write some unit tests for this scenario.

@coderabbitai, can you create a new issue that briefly describes the problem, and the suggested approach toward addressing it by refactoring and unit testing?

bielj commented on mdx2/command_line/find_peaks.py:
```diff
@@ -2,72 +2,113 @@
+    nproc = params.nproc
+
+    logger.info("Loading geometry and image data...")
+    MI = loadobj(geom, "miller_index")
+    IS = loadobj(data, "image_series")
```
From local CodeRabbit:
Add error handling for file operations.
The loadobj calls (and later saveobj calls on lines 70-72) have no error handling. If the files don't exist, are corrupted, or lack the expected data objects, the function will fail with an unclear error message.
Consider wrapping these operations in try-except blocks with informative error messages to guide users when file operations fail.
Example:
```
try:
    MI = loadobj(geom, "miller_index")
    IS = loadobj(data, "image_series")
except FileNotFoundError as e:
    logger.error(f"Data file not found: {e}")
    raise
except Exception as e:
    logger.error(f"Failed to load data: {e}")
    raise
```

spmeisburger replied:
The file handling in `loadobj` and `saveobj` is done by `nxload` and `nxsave` from the nexusformat package. In cases where the file is not found, or the name of the entry is incorrect, a NeXusError is raised. For instance, if I do:

```python
m = loadobj('sandbox/map2.nxs','intensity') # sandbox/map2.nxs does not exist
```

I get an informative error:

```
NeXusError: '/Users/steve/dev/mdx2/sandbox/map2.nxs' does not exist
```

Similarly, if the file exists but the object is not found:

```python
m = loadobj('sandbox/map.nxs','intensity2') # map.nxs exists, but intensity2 does not
```

```
NeXusError: Invalid path
```

spmeisburger replied:
It would be quite cumbersome to wrap all file io functions in try: except in order to log information. A better approach might be to check the existence & validity of input files before a program runs. We should keep this separate from parameter validation and argument parsing, since the testing framework often generates fake file names, and it would be a pain to mock. We should think more about the best way to accomplish this for the whole codebase.

@coderabbitai, can you create a new issue to address this? Specifically, we'd like a code enhancement to check the existence and validity of input files before a command-line tool executes, avoiding the creation of log files and delay between when the command is run and the issue is detected. Ideally, this would not require major changes to the command-line tools or testing framework. For instance, it could be implemented inside the `with_parsing` decorator, using some mechanism to auto-detect whether a parameter is an input file (for instance, by setting the type hint to a custom class instead of `'str'`).

bielj commented on mdx2/data.py:
```diff
@@ -285,6 +298,175 @@ def lookup(self, h, k, l, column_name, fill_value=np.nan):
+        """
+        if not isinstance(values, dict) and len(self._data) == 1:
+            k = list(self._data.keys())[0]
+            values = {k: values}
```
From local CodeRabbit:
Dead code: **kwargs always produces a dict.
The condition not isinstance(values, dict) will never be True because **kwargs always creates a dictionary, even if empty. If the intent was to allow positional array arguments, this needs a different signature.
If you intended to support passing a single array for single-column grids, consider changing the method signature:

spmeisburger replied:
This is addresed in commit [7db0729](https://github.com/diff-use/mdx2/pull/45/commits/7db0729dc51d43223b4294474057c8f754888f3f)

bielj commented on tests/test_nexus_files.py:
```diff
@@ -17,3 +28,69 @@ def test_writing_data_after_nxsave(tmp_path):
+    assert image_series.data.shape == (20, 10, 10)
+    assert len(nxobj.data._vfiles) == 3  # source files are present
+
+    # write some data to the second source file, check that it appears in the virtual dataset
+    source_series = ImageSeries.load(nxobj.data._vfiles[1], mode="r+")
+    source_series.data[0, 0, 0] = 666
+    assert image_series.data[8, 0, 0] == 666
```
From local CodeRabbit:
Close the file handle to prevent resource leaks.
The code opens a source file at line 44 but never closes it, which can cause resource leaks or test failures (especially on Windows where open file handles can prevent cleanup). Additionally, line 44 uses the private attribute _vfiles.
Apply this diff to use a context manager and ensure proper cleanup:
```
-    # write some data to the second source file, check that it appears in the virtual dataset
-    source_series = ImageSeries.load(nxobj.data._vfiles[1], mode="r+")
-    source_series.data[0, 0, 0] = 666
-    assert image_series.data[8, 0, 0] == 666
+    # write some data to the second source file, check that it appears in the virtual dataset
+    with ImageSeries.load(nxobj.data._vfiles[1], mode="r+") as source_series:
+        source_series.data[0, 0, 0] = 666
+    assert image_series.data[8, 0, 0] == 666
```

spmeisburger replied:
The file io in ImageSeries is handled by `nxload` from the `nexusformat` library. `nxload` does not leave file handles open. See previous comment: https://github.com/diff-use/mdx2/pull/45#discussion_r2615106161

bielj commented on examples/insulin-multi-crystal/3-mdx2_integrate.sh:
```diff
@@ -1,16 +1,18 @@
 #!/bin/bash
 set -e
 
+NPROC=64 # modify as needed
```
Do we want to consider attempts to deduce this for most users, and set a lower cap where no information is available?
`NPROC=$(nproc) || NPROC=$(getconf _NPROCESSORS_CONF) || NPROC=4`

spmeisburger replied:
setting `nproc=-1` would accomplish this. The parameter is passed to joblib, which interprets -1 as "use all the CPUs", -2 as "use all but 1 CPU", etc. I've updated the scripts with NPROC=-1, and will commit the changes soon once I verify that they execute without errors.

spmeisburger replied:
See commit 701d3f3d8959f8a886c05eead74fa4ca8e345cee

bielj reviewed (commented):
Added some discussion points from CodeRabbit

spmeisburger committed:
modified CLI tools to not log when arg parsing fails or help is printed, streamlined entry point creation using decorator syntax for both logging and parsing

spmeisburger committed:
Updated tests to handle equal-limits case correctly, all tests now passing (68/68)

spmeisburger committed:
removed obsolate instance check from HKLGrid.accumulate, updated docstring to be consistent with function signature

spmeisburger commented on mdx2/data.py:
```diff
@@ -468,3 +750,42 @@ def index(self, miller_index, mask=None):
+# SOME UTILITY FUNCTIONS USED ABOVE
+
+
+def _default_stack_from_rotation_angle(ang):
+    """Compute a reasonable image stack size (1--50 images) given the rotation angle per frame (degrees)
+
+    The heuristics are designed to work with typical choices of angular range per image, such as
+    0.1, 0.025, 1/3, 0.04, etc. The algorithm chooses either 40, 45, or 50 image stack size if the
+    total oscillation is less than 6 degrees. Otherwise, it chooses a stack to give either 5 or 6 degrees
+    of total oscillation. If the heuristics fail, it will make a sane choice: 10 frames for large rotation
+    per frame, and 40 frames if the rotation is small, and 50 frames if there is no rotation.
+    """
+    if ang == 0:
+        numer, denom = 0, 1
+    else:
+        numer = 180
+        denom = np.round(numer / ang).astype(int)
+    if denom < numer * 7:
+        if 5 * denom % numer == 0:
+            nimgs = 5 * denom // numer
+        elif 6 * denom % numer == 0:
+            nimgs = 6 * denom // numer
+        else:
+            nimgs = 10
+    elif 2 * 50 * numer % denom == 0:
+        nimgs = 50
+    elif 2 * 40 * numer % denom == 0:
+        nimgs = 40
+    elif 2 * 45 * numer % denom == 0:
+        nimgs = 45
+    elif 4 * 50 * numer % denom == 0:
+        nimgs = 50
+    elif 4 * 45 * numer % denom == 0:
+        nimgs = 45
+    else:
+        nimgs = 40
+    return nimgs
```
Agreed, please create an issue to address this in future.

spmeisburger committed:
ensure that at least one scaling model is enabled before running mdx2.scale

spmeisburger commented on mdx2/command_line/scale.py:
```diff
@@ -2,328 +2,294 @@
+    mca2020: bool = False
+    """shortcut for --scaling.enable True --offset.enable True --detector.enable True --absorption.enable True"""
+
+    def __post_init__(self):
+        """Validate and process parameters after initialization"""
+        # Apply mca2020 shortcut to enable all models
+        if self.mca2020:
+            self.scaling.enable = True
+            self.detector.enable = True
+            self.absorption.enable = True
+            self.offset.enable = True
+
+        # Check for duplicate input files
+        # Note: Same filename in different directories is valid (e.g., /path1/data.nxs, /path2/data.nxs)
+        # But truly duplicate paths are not allowed (e.g., data.nxs, data.nxs)
+        if len(self.hkl) != len(set(self.hkl)):
+            raise ValueError("Duplicate input files are not allowed")
+
+        # Auto-generate output file names if not provided
+        # This handles various patterns:
+        # - Different directories → scales.nxs in each directory
+        # - Same directory with underscore pattern → scales_<postfix>.nxs
+        # - Other cases return None and trigger error below
+        if self.outfile is None:
+            self.outfile = generate_default_outfiles(self.hkl)
+            if self.outfile is None:
+                raise ValueError("unable to auto-generate output file names from input name pattern")
+
+        # Validate that the number of output files matches the number of input files
+        if len(self.outfile) != len(self.hkl):
+            raise ValueError(
+                f"Number of output files ({len(self.outfile)}) must match number of input files ({len(self.hkl)})"
+            )
+
```
Calling mdx2.scale with all the scaling models disabled is an invalid use of the function. This should be caught in the `__post_init__` of the Parameters dataclass. I'll add this and commit.

spmeisburger committed:
fixed docstring typo: data --> geometry

spmeisburger committed:
in mdx2.merge, validate len(scale) matches number of HKL inputs/batches

spmeisburger committed:
removed duplicated call to make_argument_parser

spmeisburger committed:
Add safety checks in mdx2.import_data for virtual dataset correctness

spmeisburger committed:
fix a bug in dxtbx w/ cbf files that are imported without reading all the headers. In that case, only the first element in the list of exposure times is set, and the rest are zero.

spmeisburger commented on mdx2/command_line/import_data.py:
```diff
@@ -2,51 +2,91 @@
+    logger.info("Creating virtual dataset structure...")
+    image_series.save(
+        outfile,
+        virtual=True,
+        source_directory=datastore,
+    )
+
+    def write_stack(istart, istop, filename, datapath):
+        source = nxload(filename, "r+")[datapath]
+        data = iset.read_stack(istart, istop)
+        source[:, :, :] = data
+
+    slices = [sl for sl in image_series.chunk_slice_along_axis(0)]
+
+    # Access virtual dataset information through public API
+    # Reload the ImageSeries to ensure we have the updated virtual dataset
+    image_series_reloaded = ImageSeries.load(outfile)
+    files = image_series_reloaded.virtual_source_files
+    vpath = image_series_reloaded.virtual_dataset_path
+
+    # Properties will raise RuntimeError if internal API has changed
+    # If data is not a virtual dataset, they return None
+    if files is None or vpath is None:
+        raise RuntimeError(f"Expected virtual dataset in {outfile}, but virtual dataset information is not available.")
 
-    if args.chunks is not None:
-        image_series.data.chunks = tuple(args.chunks)
+    logger.info("Writing {} image batches (requested n_jobs: {})...", len(slices), nproc)
+    with Parallel(n_jobs=nproc, verbose=10) as parallel:
+        log_parallel_backend(parallel)
+        parallel(delayed(write_stack)(sl.start, sl.stop, fn, vpath) for sl, fn in zip(slices, files))
+    logger.info("Image data writing completed")
```
mdx2.import_data works fine with the loky backend, which verifies that iset is picklable. No changes needed.

Safety checks for virtual dataset implementation will be included in the next commit.

spmeisburger commented on mdx2/command_line/bin_image_series.py:
```diff
@@ -2,50 +2,114 @@
+        new_data = np.stack(parallel(delayed(binslab)(sl) for sl in sl_0))
+    logger.info("Binning computation completed")
+
+    # Convert to count rate
+    new_times = np.array([image_series.exposure_times[sl].mean() for sl in sl_0])
+    new_data = new_data / new_times[:, np.newaxis, np.newaxis]
+
```
We've encountered a bug before related to exposure times being zero -- in dxtbx, when cbfs are imported without reading all the image headers, only the first value of exposure times list is set, and the rest are zero. 

Rather than check exposure_times here, I put a check in dxtbx_machinery so this case can be resolved silently when the data are loaded in, rather than here. This will be in the next commit.

spmeisburger replied:
See commit c9a7da370df4e91dfa3a521c023ff70753aaae2f

Commit:
regenerate outputs in insulin tutorial notebooks

Commit:
set --nproc -1 to use all available CPUs by default

spmeisburger committed:
tightened up docstring for generate_default_outfiles

spmeisburger committed:
Add default factories for nested parameter groups to make programmatic construction usable

spmeisburger commented on mdx2/dxtbx_machinery.py:
```diff
@@ -19,6 +18,61 @@ def __init__(self, expt):
+            active_areas = sensor_active_areas((self._panel,))
+            if active_areas is None or len(active_areas) == 0:
+                raise RuntimeError(
+                    "Experiment detector geometry not found in dxtbx detector database; "
+                    "mdx2 may not be compatible with this detector type"
+                )
+
+            # Extract the first (and only) active area tuple
+            active_area = active_areas[0]
+            # Note: active_area is (fast_start, slow_start, fast_end, slow_end)
+            fast_start, slow_start, fast_end, slow_end = active_area
+            image_size = self._panel.get_image_size()  # (fast, slow)
+
+            # Module size is the active area dimension
+            self._module_size_slow = slow_end - slow_start
+            self._module_size_fast = fast_end - fast_start
+
+            # Calculate number of modules in each direction
+            # image_size is (fast, slow), so image_size[1] is slow, image_size[0] is fast
+            num_modules_slow = image_size[1] // self._module_size_slow
+            num_modules_fast = image_size[0] // self._module_size_fast
+
+            # Calculate gap sizes
+            if num_modules_slow > 1:
+                total_gap_slow = image_size[1] - (num_modules_slow * self._module_size_slow)
+                self._gap_slow = total_gap_slow // (num_modules_slow - 1)
+            else:
+                self._gap_slow = 0
+
+            if num_modules_fast > 1:
+                total_gap_fast = image_size[0] - (num_modules_fast * self._module_size_fast)
+                self._gap_fast = total_gap_fast // (num_modules_fast - 1)
+            else:
+                self._gap_fast = 0
+
+        except (AssertionError, RuntimeError) as e:
+            raise RuntimeError(f"Experiment detector not compatible with mdx2: {e}") from e
+        except Exception as e:
+            raise RuntimeError(f"Unexpected error checking detector compatibility: {e}") from e
+
```
This is a bit out of scope for this PR. @coderabbitai, can you generate an issue to address this comment, by first creating unit tests?

spmeisburger commented on mdx2/dxtbx_machinery.py:
```diff
@@ -362,14 +391,24 @@ def interp_rotation_matrices(U1, U2, rotation_fraction):
-        U2 = crystal.get_U_at_scan_point(b + 1)
-        Uvals = interp_rotation_matrices(U1, U2, f)
-        U[ind, :, :] = np.array(Uvals).reshape(3, 3)
+    if crystal.num_scan_points == 0:
+        U = np.array(crystal.get_U()).reshape(3, 3)
+        U = np.tile(U, (np.size(phi_vals_deg), 1, 1))
+    else:
+        base, frac = phi_to_base_fraction_index(scan, phi_vals_deg)
+        imax = crystal.num_scan_points
+        U = np.empty([np.size(base), 3, 3])
+        for ind, (b, f) in enumerate(zip(base, frac)):
+            U1 = crystal.get_U_at_scan_point(b)
+            # Handle boundary: clamp second index to valid range
+            b2 = min(b + 1, imax - 1)
+            if b2 == b:
+                # At the boundary, use U1 directly without interpolation
+                Uvals = U1
+            else:
+                U2 = crystal.get_U_at_scan_point(b2)
+                Uvals = interp_rotation_matrices(U1, U2, f)
+            U[ind, :, :] = np.array(Uvals).reshape(3, 3)
     return U
```
This section of the code has gotten rather difficult to understand, and I'm having a hard time reasoning through the bug here. I think the right approach might be to generate unit tests that cover this edge case. Overall, the dxtbx_machinery could use better testing coverage. Can you create an issue to improve dxtbx_machinery test coverage, and you can give this particular bug as an example.

spmeisburger replied:
Oh, I just noticed that you've been creating issues in the ando-lab/mdx2 github. The issues should be in the diff-use/mdx2 fork.  Are you able to move these issues?

spmeisburger replied:
Yes, please close the issues created in ando-lab/mdx2 with an explanation.

Commit:
re-ran multi-crystal example notebooks

spmeisburger commented on mdx2/command_line/scale.py:
```diff
@@ -2,328 +2,297 @@
+    outfile: Optional[List[str]] = field(default=None, nargs="*")
+    """name of the output NeXus file(s). If omitted, will attempt a sensible name such as scales.nxs"""
+    mca2020: bool = False
+    """shortcut for --scaling.enable True --offset.enable True --detector.enable True --absorption.enable True"""
+
+    def __post_init__(self):
+        """Validate and process parameters after initialization"""
+        # Apply mca2020 shortcut to enable all models
+        if self.mca2020:
+            self.scaling.enable = True
+            self.detector.enable = True
+            self.absorption.enable = True
+            self.offset.enable = True
+
+        # Check for duplicate input files
+        # Note: Same filename in different directories is valid (e.g., /path1/data.nxs, /path2/data.nxs)
+        # But truly duplicate paths are not allowed (e.g., data.nxs, data.nxs)
+        if len(self.hkl) != len(set(self.hkl)):
+            raise ValueError("Duplicate input files are not allowed")
+
+        # Auto-generate output file names if not provided
+        # This handles various patterns:
+        # - Different directories → scales.nxs in each directory
+        # - Same directory with underscore pattern → scales_<postfix>.nxs
+        # - Other cases return None and trigger error below
+        if self.outfile is None:
+            self.outfile = generate_default_outfiles(self.hkl)
+            if self.outfile is None:
+                raise ValueError("unable to auto-generate output file names from input name pattern")
+
+        # Validate that the number of output files matches the number of input files
+        if len(self.outfile) != len(self.hkl):
+            raise ValueError(
+                f"Number of output files ({len(self.outfile)}) must match number of input files ({len(self.hkl)})"
+            )
+
+        # Ensure at least one model is enabled
+        if not (self.scaling.enable or self.absorption.enable or self.detector.enable or self.offset.enable):
+            raise ValueError("At least one model must be enabled: scaling, absorption, detector, or offset.")
+
```
Let's add this to Issue #51. If input files are always Path objects, then we can make use of the Path.resolve function to put them in canonical form.

spmeisburger replied:
Yes, please add this comment to issue #51.

spmeisburger commented on mdx2/dxtbx_machinery.py:
```diff
@@ -399,13 +438,23 @@ def phi_to_base_fraction_index(scan, phi_vals_deg):
-        B1 = np.array(crystal.get_B_at_scan_point(b)).reshape([3, 3])
-        B2 = np.array(crystal.get_B_at_scan_point(b + 1)).reshape([3, 3])
-        B[ind, :, :] = B1 * (1 - f) + B2 * f  # element-wise linear interpolation
+    if crystal.num_scan_points == 0:
+        B = np.array(crystal.get_B()).reshape(3, 3)
+        B = np.tile(B, (np.size(phi_vals_deg), 1, 1))
+    else:
+        base, frac = phi_to_base_fraction_index(scan, phi_vals_deg)
+        imax = crystal.num_scan_points
+        B = np.empty([np.size(base), 3, 3])
+        for ind, (b, f) in enumerate(zip(base, frac)):
+            B1 = np.array(crystal.get_B_at_scan_point(b)).reshape([3, 3])
+            # Handle boundary: clamp second index to valid range
+            b2 = min(b + 1, imax - 1)
+            if b2 == b:
+                # At the boundary, use B1 directly without interpolation
+                B[ind, :, :] = B1
+            else:
+                B2 = np.array(crystal.get_B_at_scan_point(b2)).reshape([3, 3])
+                B[ind, :, :] = B1 * (1 - f) + B2 * f  # element-wise linear interpolation
     return B
 
```
Since this resembles another comment above concerning dxtbx_machinery, can you add it to issue #52 so that we can address it properly by adding unit tests for edge cases?

spmeisburger commented on mdx2/command_line/import_data.py:
```diff
@@ -2,51 +2,97 @@
+    files = image_series_reloaded.virtual_source_files
+    vpath = image_series_reloaded.virtual_dataset_path
+
+    # These edge cases should not happen if virtual datasets are implemented correctly, but check anyway
+    if len(files) != len(slices):
+        raise RuntimeError(f"Virtual dataset mismatch: {len(slices)=} vs {len(files)=}")
+    if len(set(files)) != len(files):
+        raise RuntimeError("Virtual_source_files contains duplicates, cannot proceed with import.")
+
+    # Properties will raise RuntimeError if internal API has changed
+    # If data is not a virtual dataset, they return None
+    if files is None or vpath is None:
+        raise RuntimeError(f"Expected virtual dataset in {outfile}, but virtual dataset information is not available.")
 
```
A few lines above, ImageSeries is saved with "virtual=True". Here, we're re-opening the same file. So there's no need to consider a case involving a non-virtual dataset.

spmeisburger committed:
safer path manipulation when preventing mdx2 folder from shadowing imports, linting
